[Tan Wei Liang] iP

wltan opened the pull request.

Commit:
Add Gradle support

wltan committed:
Basic Greet, Echo, Exit commands

wltan committed:
Rename bot from Duke to iPbot

wltan committed:
Add basic task list operations (Add, List)

wltan committed:
Add task completion status

wltan committed:
Add ability to mark tasks as done

wltan committed:
Set command strings as constants

wltan committed:
Refactor functionality to helper methods

wltan committed:
Fix Add Task to use full string representation

wltan committed:
Add different types of tasks (Todo, Event, Deadline)

wltan committed:
Add cumulative task count on add

wltan committed:
Add Automated Text UI Testing

wltan committed:
Add error handling with exceptions

wltan committed:
Allow blank commands to idle

wltan committed:
Add delete command

wltan committed:
Refactor to use enums

wltan committed:
Update text UI tests (#2)

wltan committed:
Add automatic save and load

iPbot will now save on any command and load on startup

wltan committed:
Upgrade Deadlines and Events to use LocalDate

Previously, additional information about deadlines and events were simply stored as Strings. This makes validation difficult.

By using LocalDate and parsing the string, we can enforce some validation on the structure of the data.

wltan committed:
Merge branch 'branch-Level-7'

wltan committed:
Merge branch 'branch-Level-8'

wltan committed:
Refactor core features to separate classes

wltan committed:
Fix text UI tests

wltan committed:
Add JUnit Tests

This is a minimal implementation. More will be added as time goes on.

wltan committed:
Add manifest for Jarfile

wltan committed:
Improve Javadoc coverage

wltan committed:
Add find command

wltan committed:
Miscellaneous coding standard improvements

wltan committed:
Merge branch 'branch-A-CodingStandard'

wltan committed:
Merge branch 'branch-Level-9'

wltan committed:
Fix unit test

Tests for this class may interfere with one another due to the internal TaskList state.

Let us clear the state between tests so that each test is independent.

wltan committed:
Add Gradle support for run

This allows Gradle to support both run and test tasks.

wltan committed:
Add JavaFX GUI

GUI right now closely mirrors CLI. Enhancements will be added in a future commit.

wltan committed:
Add varargs example

This change enables output to accept multiple strings at once. The strings will be joined by newlines to make one outputString.

junhui-phoon commented on src/main/java/command/Command.java:
```diff
@@ -0,0 +1,56 @@
+    FIND_CMD    ("find", false);
+
+    private final String cmdString;
+    private final boolean noArgs;
```
Maybe you could rename 'noArgs' to 'hasNoArguments' or 'hasArguments' <--[you might have to rewrite some code if you change to this :(]

xinyee20 replied:
Yup, I agree! I think 'hasNoArgs' may be easier to understand than 'isNoArgs' :)

wltan replied:
Fixed, thanks!

wltan committed:
Add Checkstyle configuration

Violations will be fixed in a separate commit.

joshualiangxy commented on src/main/java/command/Command.java:
```diff
@@ -0,0 +1,56 @@
+/**
+ * Enumerates the types of possible commands.
+ */
+public enum Command {
```
I really like how you use `enum` here to define the different commands!

joshualiangxy commented on src/main/java/command/CommandHandler.java:
```diff
@@ -0,0 +1,180 @@
+/**
+ * A function that accepts and processes lines of user commands.
+ */
+public class CommandHandler implements Consumer<String> {
```
I really like how you implemented the `Consumer` interface for this class!

wltan replied:
Thanks!

joshualiangxy commented on src/main/java/command/CommandHandler.java:
```diff
@@ -0,0 +1,180 @@
+ */
+public class CommandHandler implements Consumer<String> {
+
+    private final boolean printOutput;
```
Consider naming the boolean something like `shouldPrint` to make it sound more like a boolean?

wltan replied:
Good point, I would have fixed it but I have since modified how this class works and there is no longer a boolean attribute.

joshualiangxy commented on src/main/java/exceptions/DataException.java:
```diff
@@ -0,0 +1,26 @@
+     */
+    public DataException(String field, String reason) {
+        super(reason);
+        this.command = field;
```
Maybe you can leave out the `this` keyword when it is not needed to reduce unnecessary noise?

wltan replied:
If possible, I would prefer to keep the `this` to make it clear that I am referring to the instance variable. Does the coding standard have anything that says otherwise?

In any case, I have renamed the variable to `field` as well. The original name `command` was a mistake from copying `CommandException`.

joshualiangxy commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,55 @@
+public abstract class Task {
+
+    // task description
+    private final String desc;
```
Perhaps a more descriptive name here would help the reader understand what the string represents?

xinyee20 replied:
I agree, since 'desc' is used in quite a few places, it may be better to use the full name 'description' so that it will be clearer in other parts of your code

wltan replied:
Fixed, thanks!

joshualiangxy reviewed (commented):
It seems that most of your code already follows the coding conventions. Keep up the good work!

xinyee20 commented on src/main/java/core/TaskList.java:
```diff
@@ -0,0 +1,57 @@
+ */
+public class TaskList {
+
+    private static final ArrayList<Task> tasks = new ArrayList<>();
```
Perhaps 'tasks' should be entirely in uppercase since you marked it as a constant using 'final'?

xinyee20 replied:
Below you have a function 'add' to add Task t to the ArrayList<Task> task, should the statement still have 'final' inside then since the array list will be changed? It may be less confusing to remove the term 'final'

wltan replied:
I don't agree that `tasks` is a constant simply because of the word `final`. As you have pointed out, the `ArrayList` is still mutable.

But I do want to keep the `final` modifier because I want to make it clear that `tasks` should not be assigned to.

xinyee20 commented on src/main/java/tasks/Task.java:
```diff
@@ -0,0 +1,55 @@
+    private final String desc;
+
+    // indicates whether the task is done or not
+    private boolean done;
```
'done' here could be changed to 'isDone' to sound more like a boolean

wltan replied:
Fixed, thanks!

xinyee20 commented on src/main/java/core/Storage.java:
```diff
@@ -0,0 +1,44 @@
+                .boxed()
+                .map(i ->  {
+                    Task t = TaskList.get(i - 1);
+                    return t.getParentCommand() + (t.getDoneStatus() ? "\ndone " + i : "");
```
Perhaps you can break before the operator '+' to improve readability

wltan replied:
Ok, sounds reasonable since the second half represents a new line in the string.

wltan committed:
Improve Code Quality from PR reviews

Based on PR peer review feedback.

wltan committed:
Fix Checkstyle violations

wltan committed:
Add assertions

wltan committed:
Merge pull request #6 from wltan/branch-A-Assertions

Add assertions

wltan committed:
Merge branch 'master' into branch-A-CodeQuality

wltan committed:
Merge pull request #5 from wltan/branch-A-CodeQuality

Code Quality Improvements

wltan committed:
Add CI workflow

Sets up automated runing of unit tests and Checkstyle.

wltan committed:
Fix method naming typo

Previous method name was confusing due to an oversight in refactoring.

wltan committed:
Add help command

wltan committed:
Wrap text output in GUI

wltan committed:
Add load command

wltan committed:
Fix jar file generation

wltan committed:
Enhance GUI

Makes output uneditable, differentiates user input from bot output.
Adds a command reference on the left. Adds a task list on the right, updated after each command.

wltan committed:
Fix exit command behavior

Previously the bot did nothing other than print "Goodbye!". Now, it will exit the application.

wltan committed:
Set up user guide

wltan committed:
Update README.md
